businessRules/donations.js

```diff
@@ -1,54 +1,59 @@
 const { v4 } = require('uuid');
 const { createDonation } = require('../models/donations');
-const { getAllIdeas } = require('../models/ideas');
-const { getIdeas } = require('../utils/storage');
+const { getIdea, validateIdea } = require('../validations/ideas');
+const { validateSum } = require('../validations/donations.js');
+const { getUser } = require('../validations/users');
 
-function createDonation(firstName, sum, userId, ideaId) {
-    let ideas = getIdeas();
-    const idea = ideas.find((idea) => ideaId === idea.id);
+function createDonationByUnregisteredUser(sum, userId, ideaId) {
+    validateIdea(ideaId);
+    validateSum(sum);
 
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
+    const user = getUser(userId);
+
+    let firstName;
+
+    if (!user) {
+        const generateId = v4();
+        const anonymous = 'Anonymous';
+
+        firstName = firstName || anonymous;
+        userId = userId || generateId;
     }
+    createDonation(firstName, sum, userId, ideaId);
+}
 
-    const generateId = v4();
-    const anonymous = 'Anonymous';
+function createDonationByRegisteredUser(sum, userId, ideaId) {
+    validateSum(sum);
 
-    firstName = firstName || anonymous;
-    userId = userId || generateId;
+    const user = getUser(userId);
+
+    const firstName = user.firstName;
 
     createDonation(firstName, sum, userId, ideaId);
 }
 
-function ideasDonationSum(ideaId) {
-    const ideas = getAllIdeas();
-    const idea = ideas.find((idea) => idea.id === ideaId);
+function getTotalSumDonatedForIdea(ideaId) {
+    const idea = getIdea(ideaId);
 
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
-    }
     return idea.totalDonationSum;
 }
 
-function ideasSumDifference(ideaId) {
-    const ideas = getAllIdeas();
-    const idea = ideas.find((idea) => idea.id === ideaId);
-
-    if (!idea) {
-        throw new Error('Error. No idea with such ID found.');
-    }
+function getIdeasSumDifference(ideaId) {
+    const idea = getIdea(ideaId);
 
     const ideasSum = idea.askedSum;
 
-    const totalDonationSumForThisIdea = ideasDonationSum(ideaId);
+    const totalDonationSumForThisIdea = getTotalSumDonatedForIdea(ideaId);
+
     const askedSumAndDonationSumDifference =
         ideasSum - totalDonationSumForThisIdea;
 
     return askedSumAndDonationSumDifference;
 }
 
 module.exports = {
-    createDonation,
-    ideasDonationSum,
-    ideasSumDifference,
+    createDonationByUnregisteredUser,
+    createDonationByRegisteredUser,
+    getTotalSumDonatedForIdea,
+    getIdeasSumDifference,
```

Comment:
The `getIdeasSumDifference` name is not clear. Can you think of a more descriptive name?

Reply:
I tryed to. Maybe new one will sound better.